### PR TITLE
feature/voyager-config-upload-path

### PR DESCRIPTION
**Pull request by shakalaka**
Took out in voyager configuration upload path

### CODE REVIEW

**Review by marktopper — DISMISSED** [review]

**Comment by marktopper** [comment]
Does it work for Windows systems? Or does it need to use `DIRECTORY_SEPARATOR`?

File: src/Helpers/helperUploadPath.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
+<?php
+
+if (!function_exists('getUploadPath')) {
```
Comment (fletch3555) [inline src/Helpers/helperUploadPath.php:3]: I think the helper method is too specific, and would be better served as a service class method or something.

File: src/Http/Controllers/ContentTypes/File.php [hunk 2]
```diff
@@ -43,7 +43,7 @@ public function handle()
      */
     protected function generatePath()
     {
-        return $this->slug.DIRECTORY_SEPARATOR.date('FY').DIRECTORY_SEPARATOR;
+        return getUploadPath($this->slug, config('voyager.storage.path', '%slug%/%date:F%/%date:Y%/'));
```
Comment (fletch3555) [inline src/Http/Controllers/ContentTypes/File.php:46]: As @marktopper said, you changed all of the DIRECTORY_SEPARATOR calls to literal strings, which very likely breaks OS cross-compatibility.

File: src/Helpers/helperUploadPath.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+        foreach ($_dirs as $dir) {
+            if ('' === $dir) {
+                continue;
+            }
+
+            if (preg_match('/%(.*?)%/', $dir)) {
+                $dir = str_replace('%', '', $dir);
+
+                if ($dir === 'slug') {
```
Comment (fletch3555) [inline src/Helpers/helperUploadPath.php:26]: I'm also not really a huge fan of the placeholder replacement.  This if/elseif/etc chain could easily get VERY long and unmanageable.

File: src/Helpers/helperUploadPath.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+     * Return upload path.
+     *
+     * @param string $slug
+     * @param string $path
+     *
+     * @return string
+     */
+    function getUploadPath($slug, $path)
+    {
+        $_dirs = explode('/', $path);
```
Comment (fletch3555) [inline src/Helpers/helperUploadPath.php:15]: Variable naming is too generic.  Also, $_ is generally reserved for PHP superglobals.

**Review by fletch3555 — CHANGES_REQUESTED** [review]
I think this could be better served by just letting Laravel/Voyager upload the files to a single place.  You can then either move them around manually, or choose a subfolder in the controller of a form using the file/image upload field.

I'm certainly not opposed to letting this be configurable, but I don't think the Voyager config file is the right place.  This would be better configured in the JSON "additional options" field, which allows it to be different for each BREAD type, and keeps the usage localized to the "ContentTypes" classes

**Comment by marktopper** [comment]
Closed due to inactivity

**Pull request closed without merging**